File: memory-gateway/app/models/memory.py
"""Canonical memory schema: types, statuses, scores, and candidates."""

from __future__ import annotations

from enum import Enum
from typing import Sequence

from pydantic import BaseModel, Field, field_validator

from app.memory.correction import Correction
from app.memory.revocation import Revocation


class MemoryType(str, Enum):
    FACT = "fact"
    DECISION = "decision"
    CONSTRAINT = "constraint"
    PREFERENCE = "preference"
    GOAL = "goal"
    ARCHITECTURE = "architecture"
    IMPORTANT_EVENT = "important_event"
    ACTIVE_TASK = "active_task"


CANONICAL_TYPES: tuple[MemoryType, ...] = tuple(MemoryType)


class MemoryStatus(str, Enum):
    ACTIVE = "active"
    SUPERSEDED = "superseded"
    REVOKED = "revoked"
    EXPIRED = "expired"
    OBSOLETE = "obsolete"


class MemoryScores(BaseModel):
    """Separate score axes — never collapse into one metric."""

    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    importance: float = Field(default=0.0, ge=0.0, le=1.0)
    stability: float = Field(default=0.0, ge=0.0, le=1.0)
    freshness: float = Field(default=1.0, ge=0.0, le=1.0)
    information_gain: float = Field(default=0.0, ge=0.0, le=1.0)


class StructuredFact(BaseModel):
    """Structured representation (entity, attribute, value) of an extracted fact."""

    entity: str
    attribute: str
    value: str
    memory_type: MemoryType = MemoryType.FACT
    raw_text: str = ""

    def to_content(self) -> str:
        if self.memory_type == MemoryType.PREFERENCE:
            return self.value
        if self.attribute == "project" and self.entity == self.value:
            return f"Project = {self.value}"
        if self.entity and self.entity.lower() not in ("user", self.attribute.lower()):
            return f"{self.entity} {self.attribute} = {self.value}"
        attr_title = self.attribute[0].upper() + self.attribute[1:] if self.attribute else "Item"
        return f"{attr_title} = {self.value}"

    def to_topic_key(self) -> str:
        from app.memory.facts import detect_preference_domain, slugify

        if self.memory_type == MemoryType.PREFERENCE:
            _, topic = detect_preference_domain(self.value)
            return topic
        if self.attribute == "project":
            return "project"
        if self.entity and self.entity.lower() not in ("user", self.attribute.lower()):
            return f"{slugify(self.entity)}_{slugify(self.attribute)}"
        return slugify(self.attribute)


class CandidateMemory(BaseModel):
    """In-memory candidate before persistence / merge."""

    content: str
    type: MemoryType
    scores: MemoryScores = Field(default_factory=MemoryScores)
    source_message_ids: list[str] = Field(default_factory=list)
    topic_key: str = ""
    authority: str = "user"  # user | assistant | speculation
    is_correction: bool = False
    structured_fact: StructuredFact | None = None
    correction: "Correction | None" = None
    revocation: "Revocation | None" = None

    @field_validator("content")
    @classmethod
    def _strip_content(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("content must be non-empty")
        return text


class CanonicalMemorySnapshot(BaseModel):
    """Grouped active memory for versioned context state."""

    facts: list[str] = Field(default_factory=list)
    decisions: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    preferences: list[str] = Field(default_factory=list)
    goals: list[str] = Field(default_factory=list)
    architecture: list[str] = Field(default_factory=list)
    important_events: list[str] = Field(default_factory=list)
    active_tasks: list[str] = Field(default_factory=list)

    @classmethod
    def from_items(cls, items: Sequence[object]) -> CanonicalMemorySnapshot:
        field_for = {
            MemoryType.FACT: "facts",
            MemoryType.DECISION: "decisions",
            MemoryType.CONSTRAINT: "constraints",
            MemoryType.PREFERENCE: "preferences",
            MemoryType.GOAL: "goals",
            MemoryType.ARCHITECTURE: "architecture",
            MemoryType.IMPORTANT_EVENT: "important_events",
            MemoryType.ACTIVE_TASK: "active_tasks",
        }
        data: dict[str, list[str]] = {name: [] for name in field_for.values()}
        for item in items:
            type_raw = getattr(item, "type", None)
            content = getattr(item, "content", None)
            if not isinstance(content, str) or not content:
                continue
            try:
                mtype = MemoryType(type_raw) if not isinstance(type_raw, MemoryType) else type_raw
            except ValueError:
                continue
            data[field_for[mtype]].append(content)
        return cls(**data)


class MemoryAIOutput(BaseModel):
    """Structured JSON contract expected from Memory AI compressor."""

    summary: str = ""
    facts: list[str] = Field(default_factory=list)
    decisions: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    preferences: list[str] = Field(default_factory=list)
    goals: list[str] = Field(default_factory=list)
    architecture: list[str] = Field(default_factory=list)
    important_events: list[str] = Field(default_factory=list)
    active_tasks: list[str] = Field(default_factory=list)
    obsolete_items: list[str] = Field(default_factory=list)
    contradictions: list[str] = Field(default_factory=list)
    confidence: float = Field(default=0.9, ge=0.0, le=1.0)


class ToolSummaryOutput(BaseModel):
    """Structured JSON contract for tool output compression."""

    summary: str
    key_points: list[str] = Field(default_factory=list)
    status: str = "success"

File: src/models/memory.ts
/**
 * Canonical memory schema: types, statuses, scores, and candidates.
 */

export enum MemoryType {
  FACT = "fact",
  DECISION = "decision",
  CONSTRAINT = "constraint",
  PREFERENCE = "preference",
  GOAL = "goal",
  ARCHITECTURE = "architecture",
  IMPORTANT_EVENT = "important_event",
  ACTIVE_TASK = "active_task",
  TEMPORARY_STATE = "temporary_state",
}

export const CANONICAL_TYPES: MemoryType[] = Object.values(MemoryType);

export enum MemoryStatus {
  ACTIVE = "active",
  SUPERSEDED = "superseded",
  REVOKED = "revoked",
  EXPIRED = "expired",
  OBSOLETE = "obsolete",
}

/**
 * First-class relationship kinds between memory items. Recorded on the
 * `relationship` column (the relation this item holds toward `supersedesId`)
 * and in the JSON id-list columns (`contradictsIdsJson`,
 * `relatedMemoryIdsJson`).
 */
export enum MemoryRelationship {
  SUPPORTS = "supports",
  CONTRADICTS = "contradicts",
  SUPERSEDES = "supersedes",
  DERIVED_FROM = "derived_from",
  RELATED_TO = "related_to",
  REVOKES = "revokes",
}

export interface MemoryScores {
  confidence: number;
  importance: number;
  stability: number;
  freshness: number;
  informationGain: number;
}

export function defaultMemoryScores(): MemoryScores {
  return {
    confidence: 0.0,
    importance: 0.0,
    stability: 0.0,
    freshness: 1.0,
    informationGain: 0.0,
  };
}

export type FactScope = "user" | "project" | "session";

/**
 * Temporal / modality state of a fact. "usage/current" when omitted for
 * backwards compatibility with facts that have no explicit state.
 */
export type FactState =
  | "current"
  | "past"
  | "planned"
  | "possible"
  | "conditional"
  | "stopped"
  | "superseded";

/** Three-way classification from the Memory Analyzer. */
export enum MemoryBucket {
  STORE = "store",
  CONTEXT = "context",
  DISCARD = "discard",
}

/** Short-term context entry stored in Redis / ephemeral store. */
export interface ContextEntry {
  key: string;
  value: string;
  /** Seconds from write time before the entry expires. 0 = no expiry. */
  ttlSeconds?: number;
  createdAt?: string;
}

export interface StructuredFact {
  entity: string;
  attribute: string;
  value: string;
  memoryType: MemoryType;
  rawText: string;
  /** Stable structured key such as "user.name" or "project.database". */
  key?: string;
  /** USER / PROJECT / SESSION. Derived from the key prefix when not set. */
  scope?: FactScope;
  /** True when this fact is an explicit update of a previous value. */
  isUpdate?: boolean;
  /** Temporal / modality state (current, past, planned, possible, …). */
  state?: FactState;
  /** Local extraction confidence 0..1 (undefined → treat as implicit). */
  confidence?: number;
  /** Optional metadata: comparison target ("over"), causal reason, etc. */
  metadata?: Record<string, string>;
}

export interface Correction {
  target: string;
  oldValue: string;
  newValue: string;
}

export interface Revocation {
  target: string;
  value: string;
}

export interface CandidateMemory {
  content: string;
  type: MemoryType;
  scores: MemoryScores;
  sourceMessageIds: string[];
  topicKey: string;
  authority: "user" | "assistant" | "speculation";
  isCorrection: boolean;
  /** Classification decided by the Memory Analyzer. */
  bucket?: MemoryBucket;
  subject?: string;
  predicate?: string;
  value?: string;
  scope?: FactScope;
  /** ISO timestamp of when this candidate becomes/claims to be valid. */
  validFrom?: string | null;
  /** ISO timestamp after which this candidate should no longer be treated as current. */
  validUntil?: string | null;
  structuredFact?: StructuredFact | null;
  correction?: Correction | null;
  revocation?: Revocation | null;
  /** Relationship this candidate holds toward an existing memory item. */
  relationship?: MemoryRelationship | null;
  /** Id of the memory item this candidate is derived from / related to. */
  relatedToId?: number | null;
}
export interface CanonicalMemorySnapshot {
  facts: string[];
  decisions: string[];
  constraints: string[];
  preferences: string[];
  goals: string[];
  architecture: string[];
  importantEvents: string[];
  activeTasks: string[];
}

export function createEmptySnapshot(): CanonicalMemorySnapshot {
  return {
    facts: [],
    decisions: [],
    constraints: [],
    preferences: [],
    goals: [],
    architecture: [],
    importantEvents: [],
    activeTasks: [],
  };
}

export function snapshotFromItems(items: Array<{ type: string; content: string }>): CanonicalMemorySnapshot {
  const snapshot = createEmptySnapshot();
  for (const item of items) {
    if (!item.content) continue;
    switch (item.type) {
      case MemoryType.FACT:
        snapshot.facts.push(item.content);
        break;
      case MemoryType.DECISION:
        snapshot.decisions.push(item.content);
        break;
      case MemoryType.CONSTRAINT:
        snapshot.constraints.push(item.content);
        break;
      case MemoryType.PREFERENCE:
        snapshot.preferences.push(item.content);
        break;
      case MemoryType.GOAL:
        snapshot.goals.push(item.content);
        break;
      case MemoryType.ARCHITECTURE:
        snapshot.architecture.push(item.content);
        break;
      case MemoryType.IMPORTANT_EVENT:
        snapshot.importantEvents.push(item.content);
        break;
      case MemoryType.ACTIVE_TASK:
        snapshot.activeTasks.push(item.content);
        break;
      case MemoryType.TEMPORARY_STATE:
        break;
    }
  }
  return snapshot;
}

export interface MemoryAICandidate {
  action: "NEW" | "REINFORCE" | "SUPERSEDES" | "CONTRADICTS" | "UPDATE" | "DISCARD";
  destination: "STORE" | "CONTEXT" | "DISCARD";
  type: string;
  scope: "USER" | "PROJECT" | "SESSION";
  subject: string;
  predicate: string;
  value: string;
  topicKey: string;
  confidence: number;
  importance: number;
  stability: "permanent" | "long-term" | "short-term" | "session";
  ttl_hours: number | null;
  supersedes_id: string | null;
  reinforces_id: string | null;
  informationGain: number;
  rawText: string;
}

/** Validated Memory AI extraction result (array of structured candidates). */
export interface MemoryAIOutput {
  candidates: MemoryAICandidate[];
  /** Derived from SUPERSEDES / CONTRADICTS for engine obsolete marking. */
  obsolete_items: string[];
}

/** One consolidated memory produced by the consolidation engine. */
export interface ConsolidatedMemory {
  type: string;
  scope: "USER" | "PROJECT" | "SESSION";
  subject: string;
  predicate: string;
  value: string | Record<string, unknown>;
  topicKey: string;
  confidence: number;
  importance: number;
  stability: "permanent" | "long-term" | "short-term" | "session";
  sourceMemoryIds: string[];
  consolidationNote?: string;
}

export interface ConsolidationConflict {
  memory_ids: string[];
  description: string;
  resolution: string;
}

/** Result of consolidating one related memory cluster. */
export interface ConsolidationResult {
  consolidated: ConsolidatedMemory[];
  superseded_ids: string[];
  conflicts_detected: ConsolidationConflict[];
}

export interface ToolSummaryOutput {
  summary: string;
  key_points: string[];
  status: "success" | "error" | "warning";
}
